marklogic-client-api/src/main/java/com/marklogic/client/datamovement/JSONSplitter.java

```diff
@@ -0,0 +1,550 @@
+            arrayDepth++;
+
+            if (arrayDepth > 1) {
+                arrayDepth--;
+                return NodeOperation.PROCESS;
+            }
+
+            return NodeOperation.DESCEND;
+        }
+
+        /**
+         * Receives a notification when hitting end of array, and decreases arrayDepth.
+         * @param containerKey The key of the object which the value contains current array
+         */
+        public void endArray(String containerKey) {
+            arrayDepth--;
+        }
+
+        /**
+         * Construct buffered StringHandles from JsonParser.
+         * @param containerParser the JsonParser with target object or array
+         * @return the StringHandle with target object or array
+         */
+        public StringHandle makeBufferedHandle(JsonParser containerParser) {
+            if (containerParser == null) {
+                throw new IllegalArgumentException("JsonParser cannot be null");
+            }
+            String content = serialize(containerParser);
+            return new StringHandle(content).withFormat(Format.JSON);
+        }
+    }
+
+    private static abstract class JSONSpliterator<U, T extends AbstractWriteHandle> extends Spliterators.AbstractSpliterator<U> {
+
+        private JsonParser jsonParser;
+
+        private JsonParser getJsonParser() {
+            return this.jsonParser;
+        }
+
+        private void setJsonParser(JsonParser jsonParser) {
+            this.jsonParser = jsonParser;
+        }
+
+        private ArrayDeque<String> key = new ArrayDeque<>();
+        private Visitor<T> visitor;
+        private JSONSplitter<T> splitter;
+
+        private void setSplitter(JSONSplitter<T> splitter) {
+            if (splitter == null) {
+                throw new IllegalArgumentException("JSONSplitter cannot be null");
+            }
+            this.splitter = splitter;
+        }
+
+        JSONSplitter<T> getSplitter() {
+            return this.splitter;
+        }
+
+        JSONSpliterator(JSONSplitter<T> splitter, JsonParser jsonParser) {
+            super(Long.MAX_VALUE, Spliterator.NONNULL + Spliterator.IMMUTABLE);
+            setSplitter(splitter);
+            setJsonParser(jsonParser);
+            this.visitor = splitter.getVisitor();
+        }
+
+        T getNextHandle() {
+            try {
+                while (jsonParser.nextToken() != null) {
+                    JsonToken currentToken = jsonParser.currentToken();
+
+                    switch (currentToken) {
+                        case FIELD_NAME:
+                            key.pop();
+                            key.push(jsonParser.getCurrentName());
+                            break;
+
+                        case START_OBJECT:
+                            NodeOperation operation = visitor.startObject(key.peek());
+                            switch (operation) {
+                                case DESCEND:
+                                    key.push(currentToken.asString());
```

Comment:
Should the START_OBJECT case always push a string by moving line 351 before line 348, given that the END_OBJECT case always pops a key?

Reply:
No, because only in case DESCEND, we go down into this object and will reach END_OBJECT. In two other cases, PROCESS and SKIP, we won't reach END_OBJECT to pop out the key.